packages/flutter/lib/src/widgets/scrollable_grid.dart

```diff
@@ -0,0 +1,219 @@
+  final GridDelegate delegate;
+  final ExtentsChangedCallback onExtentsChanged;
+  final List<Widget> children;
+
+  RenderGrid createRenderObject() => new RenderGrid(delegate: delegate);
+
+  _GridViewportElement createElement() => new _GridViewportElement(this);
+}
+
+// TODO(abarth): This function should go somewhere more general.
+int _lowerBound(List sortedList, var value, { int begin: 0 }) {
+  int current = begin;
+  int count = sortedList.length - current;
+  while (count > 0) {
+    int step = count >> 1;
+    int test = current + step;
+    if (sortedList[test] < value) {
+      current = test + 1;
+      count -= step + 1;
+    } else {
+      count = step;
+    }
+  }
+  return current;
+}
+
+class _GridViewportElement extends RenderObjectElement<GridViewport> {
+  _GridViewportElement(GridViewport widget) : super(widget);
+
+  double _contentExtent;
+  double _containerExtent;
+
+  int _materializedChildBase;
+  int _materializedChildCount;
+
+  List<Element> _materializedChildren = const <Element>[];
+
+  GridSpecification _specification;
+  double _repaintOffsetBase;
+  double _repaintOffsetLimit;
+
+  RenderGrid get renderObject => super.renderObject;
+
+  void visitChildren(ElementVisitor visitor) {
+    if (_materializedChildren == null)
+      return;
+    for (Element child in _materializedChildren)
+      visitor(child);
+  }
+
+  void mount(Element parent, dynamic newSlot) {
+    super.mount(parent, newSlot);
+    renderObject.callback = layout;
+    _updateRenderObject();
+  }
+
+  void unmount() {
+    renderObject.callback = null;
+    super.unmount();
+  }
+
+  void update(GridViewport newWidget) {
+    super.update(newWidget);
+    _updateRenderObject();
+    if (!renderObject.needsLayout)
+      _updateChildren();
+  }
+
+  void _updatePaintOffset() {
+    renderObject.paintOffset = new Offset(0.0, -(widget.startOffset - _repaintOffsetBase));
+  }
+
+  void _updateRenderObject() {
+    renderObject.delegate = widget.delegate;
+    renderObject.virtualChildCount = widget.children.length;
+
+    if (_specification != null) {
+      _updatePaintOffset();
+
+      // If we don't already need layout, we need to request a layout if the
+      // viewport has shifted to expose a new row.
+      if (!renderObject.needsLayout) {
+        if (_repaintOffsetBase != null && widget.startOffset < _repaintOffsetBase)
+          renderObject.markNeedsLayout();
+        else if (_repaintOffsetLimit != null && widget.startOffset + _containerExtent > _repaintOffsetLimit)
+          renderObject.markNeedsLayout();
+      }
+    }
+  }
+
+  void _updateChildren() {
+    assert(_materializedChildBase != null);
+    assert(_materializedChildCount != null);
+    List<Widget> newWidgets = new List<Widget>(_materializedChildCount);
+    for (int i = 0; i < _materializedChildCount; ++i) {
+      Widget child = widget.children[_materializedChildBase + i];
+      newWidgets[i] = new RepaintBoundary(key: new ObjectKey(child), child: child);
```

Comment:
not sure how you're gonna do this, but you can't use the child's identity or the index to be the key here.